File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let getValue die =
    match die with
    | One -> 1
    | Two -> 2
    | Three -> 3
    | Four -> 4
    | Five -> 5
    | Six -> 6

let checkIfStraight dice initial = 
    dice 
    |> List.map (fun die -> getValue die) 
    |> List.sort
    |> List.fold ( fun (last, valid) current -> 
        if valid && current = last + 1 then (current, true) else (current,false) ) (initial,true)
    |> snd

let getGroups dice =
    dice 
    |> List.groupBy id 
    |> List.map(fun (key, elements)->  (key, elements |> Seq.length) ) 

let checkIfSingular dice = dice |> getGroups |> List.length = 1

let getSingleScore die dice =
    let tuple = getGroups dice |> List.filter( fun tuple -> (fst tuple) = die ) |> List.tryHead
    match tuple with
    | None -> 0
    | Some(x) -> 
        let (_, count) = x
        count * getValue die

let score category dice = 
    match category with 
    | Ones -> getSingleScore One dice
    | Twos -> getSingleScore Two dice
    | Threes -> getSingleScore Three dice
    | Fours -> getSingleScore Four dice
    | Fives -> getSingleScore Five dice
    | Sixes -> getSingleScore Six dice
    | FullHouse -> 
        let groups =getGroups dice |> List.map (fun (die,count) ->(getValue die, count) ) |> List.sortBy fst
        match List.length groups, List.head groups |> snd with
        | count, cardinality when count = 2 && cardinality > 1 && cardinality < 4-> groups |> List.map (fun (value,count) -> value * count) |> List.sum
        | _ -> 0
    | FourOfAKind -> 
        match dice |> getGroups |> List.filter (fun (_,count) -> count > 3) |> List.sortByDescending snd with
        | (die,_) :: _  -> getValue die * 4 
        | _ -> 0
    | LittleStraight -> if checkIfStraight dice 0 then 30 else 0
    | BigStraight -> if checkIfStraight dice 1 then 30 else 0
    | Yacht -> if checkIfSingular dice then 50 else 0
    | Choice -> dice |> List.map getValue |> List.sum
    | _ -> 0

File: python/yacht/yacht.py
"""
This exercise stub and the test suite contain several enumerated constants.

Since Python 2 does not have the enum module, the idiomatic way to write
enumerated constants has traditionally been a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it’s memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""


# Score categories.
# Change the values as you see fit.

YACHT = "Y"
ONES = "1"
TWOS = "2"
THREES = "3"
FOURS = "4"
FIVES = "5"
SIXES = "6"
FULL_HOUSE = "FH"
FOUR_OF_A_KIND = "F"
LITTLE_STRAIGHT = "LS"
BIG_STRAIGHT = "BS"
CHOICE = "C"


def score(dice, category):
    """
    :param dice:
    :param category:
    :return:
    """

    if category == YACHT:
        result = 50 if len(set(dice)) == 1 else 0
    elif category == ONES:
        result = score_number(dice,1)
    elif category == TWOS:
        result = score_number(dice,2)
    elif category== THREES:
        result = score_number(dice,3)
    elif category== FOURS:
        result = score_number(dice,4)
    elif category== FIVES:
        result = score_number(dice,5)
    elif category== SIXES:
        result = score_number(dice,6)
    elif category== FULL_HOUSE:
        result = score_full_house(dice)
    elif category== FOUR_OF_A_KIND:
        result = score_four_of_a_kind(dice)
    elif category== LITTLE_STRAIGHT:
        result = 0 if set(dice) != set(range(1,6)) else 30
    elif category== BIG_STRAIGHT:
        result = 0 if set(dice) != set(range(2,7)) else 30
    else:
        result = sum(dice)

    return result

def score_number(dice, number):
    """
    :param dice:
    :return:
    """
    return dice.count(number) * number

def score_full_house(dice):
    """
    :param dice:
    :return:
    """
    dice_set = [(d,dice.count(d)) for d in set(dice)]
    return 0 if len(dice_set) != 2 or {dice_set[0][1], dice_set[1][1]} != set([2,3]) else sum(dice)

def score_four_of_a_kind(dice):
    """
    :param dice:
    :return:
    """

    repeats = [(d, dice.count(d)) for d in dice[0:2] if dice.count(d) > 1]
    repeats.sort( key = lambda e : e[1],reverse=True)

    return 0 if len(repeats) == 0 or repeats[0][1] < 4 else 4 * repeats[0][0]
